app/src/main/java/org/fossasia/openevent/general/event/EventsDataSource.kt

```diff
@@ -0,0 +1,58 @@
+package org.fossasia.openevent.general.event
+
+import androidx.lifecycle.MutableLiveData
+import androidx.paging.PageKeyedDataSource
+import io.reactivex.disposables.CompositeDisposable
+import io.reactivex.rxkotlin.plusAssign
+import org.fossasia.openevent.general.utils.extensions.withDefaultSchedulers
+
+class EventsDataSource(
+    private val eventService: EventService,
+    private val compositeDisposable: CompositeDisposable,
+    private val query: String?,
+    private val mutableProgress: MutableLiveData<Boolean>
+
+    ) : PageKeyedDataSource<Int, Event>() {
+    override fun loadInitial(
+        params: LoadInitialParams<Int>,
+        callback: LoadInitialCallback<Int, Event>
+    ) {
+        createObservable(1, 2, callback, null)
+    }
+
+    override fun loadAfter(params: LoadParams<Int>, callback: LoadCallback<Int, Event>) {
+        val page = params.key
+        createObservable(page, page + 1, null, callback)
+    }
+
+    override fun loadBefore(params: LoadParams<Int>, callback: LoadCallback<Int, Event>) {
+        val page = params.key
+        createObservable(page, page - 1, null, callback)
+    }
+
+    fun getProgressLiveStatus(): MutableLiveData<Boolean> {
+        return mutableProgress
+    }
+
+    private fun createObservable(
+        requestedPage: Int,
+        adjacentPage: Int,
+        initialCallback: LoadInitialCallback<Int, Event>?,
+        callback: LoadCallback<Int, Event>?
+    ) {
+        compositeDisposable +=
+            eventService.getEventsByLocationPaged(query, requestedPage)
+                .withDefaultSchedulers()
+                .doFinally {
+                    mutableProgress.value = false
```

Comment:
would this approach be okay? i couldn't find any other method to get this working @iamareebjamal

Reply:
What is `getEventsByLocationPaged` returning?

Reply:
It is returning a Observable Paged List

Reply:
You cannot use doFinally here, move this call inside subscribe

Reply:
Okay but inside the individual call of page finally is working

Reply:
Should i add a check using using pageNo if equal pageNo   == 1 then update the value because we want to hide the shimmer only once

Reply:
Yes, that's reasonable